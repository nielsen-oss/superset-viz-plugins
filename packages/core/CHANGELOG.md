# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 0.1.1 (2020-12-30)

**Note:** Version bump only for package @superset-viz-plugins/core





# 0.1.0 (2020-12-28)


### Features

* basic core package lib ([5f30462](https://github.com/nielsen-oss/superset-viz-plugins/commit/5f304621fbeea6554740eb2bfd156338f772429e))
