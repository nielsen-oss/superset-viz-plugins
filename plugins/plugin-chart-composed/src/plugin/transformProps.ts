/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { ChartProps } from '@superset-ui/core';
import { AxisInterval } from 'recharts';
import { CHART_SUB_TYPES, CHART_TYPES, Layout, mergeBy } from '../components/utils';
import { ComposedChartProps } from '../components/ComposedChart';
import {
  addBreakdownYColumnsAndGetBreakdownValues,
  addRechartsKeyAndGetXColumnValues,
  checkTimeSeries,
  Data,
  FormData,
  getChartSubType,
  processNumbers,
  QueryMode,
  ResultData,
  SortingType,
  sortOrderedBars,
} from './utils';

export default function transformProps(chartProps: ChartProps) {
  const { width, height, queriesData } = chartProps;
  const data = queriesData[0].data as Data[];
  const formData = chartProps.formData as FormData;

  let xColumns: string[];
  let yColumns: string[];

  if (formData.queryMode === QueryMode.raw) {
    xColumns = [formData.xColumn];
    yColumns = [formData.yColumn];
  } else {
    xColumns = formData.groupby;
    yColumns = formData.metrics?.map(metric => metric.label);
  }

  const xColumnValues: string[] = [];
  const isTimeSeries = checkTimeSeries(xColumns, formData.granularitySqla, formData.layout);
  let resultData: ResultData[] = addRechartsKeyAndGetXColumnValues(
    formData,
    data,
    xColumnValues,
    isTimeSeries,
    xColumns,
  );

  if (isTimeSeries) {
    resultData.sort(
      ({ __timestamp: __timestamp2 }, { __timestamp: __timestamp1 }) =>
        (__timestamp2 as number) - (__timestamp1 as number),
    );
  }

  const breakdowns: string[] = [];
  resultData = addBreakdownYColumnsAndGetBreakdownValues(resultData, yColumns, formData, breakdowns);

  // Unit data elements by groupBy values
  resultData = mergeBy(resultData, 'rechartsDataKey');

  const chartSubType = getChartSubType(
    formData.chartType,
    formData.barChartSubType,
    formData.lineChartSubType,
    formData.areaChartSubType,
    formData.scatterChartSubType,
  );

  const chartTypeMetrics: (keyof typeof CHART_TYPES)[] = [];
  const chartSubTypeMetrics: (keyof typeof CHART_SUB_TYPES)[] = [];
  const useCustomTypeMetrics: boolean[] = [];

  if (formData.queryMode !== QueryMode.raw) {
    yColumns.forEach((yColumn, index) => {
      useCustomTypeMetrics.push(formData[`useCustomTypeMetric${index}`] as boolean);
      chartTypeMetrics.push(formData[`chartTypeMetric${index}`] as keyof typeof CHART_TYPES);
      chartSubTypeMetrics.push(
        getChartSubType(
          formData[`chartTypeMetric${index}`] as keyof typeof CHART_TYPES,
          formData[`barChartSubTypeMetric${index}`] as keyof typeof CHART_SUB_TYPES,
          formData[`lineChartSubTypeMetric${index}`] as keyof typeof CHART_SUB_TYPES,
          formData[`areaChartSubTypeMetric${index}`] as keyof typeof CHART_SUB_TYPES,
          formData[`scatterChartSubTypeMetric${index}`] as keyof typeof CHART_SUB_TYPES,
        ),
      );
    });
  }

  let resultShowTotals = false;
  if (
    formData.barChartSubType === CHART_SUB_TYPES.STACKED &&
    formData.chartType === CHART_TYPES.BAR_CHART &&
    useCustomTypeMetrics.every(el => !el)
  ) {
    resultShowTotals = formData.showTotals;
  }

  const orderByYColumn = formData.useOrderByMetric0 && formData.orderByTypeMetric0;

  const hasOrderedBars = formData.chartType === CHART_TYPES.BAR_CHART && !!orderByYColumn;

  if (hasOrderedBars) {
    sortOrderedBars(
      resultData,
      xColumnValues,
      formData,
      formData.queryMode === QueryMode.raw ? 'XAxisColumn' : 'GroupBy',
    );
  }

  resultData = processNumbers(resultData, breakdowns, formData.numbersFormat, formData.numbersFormatDigits);
  const result: ComposedChartProps = {
    orderByYColumn: orderByYColumn as SortingType,
    hasOrderedBars,
    minBarWidth: formData.minBarWidth,
    breakdowns,
    width,
    height,
    isTimeSeries,
    xColumns,
    yColumns,
    chartTypeMetrics,
    chartSubTypeMetrics,
    hasCustomTypeMetrics: useCustomTypeMetrics,
    layout: formData.layout,
    colorScheme: formData.colorScheme,
    chartType: formData.chartType,
    showLegend: formData.showLegend,
    legendPosition: formData.legendPosition,
    chartSubType,
    showTotals: resultShowTotals,
    numbersFormat: formData.numbersFormat,
    labelsColor: formData.labelsColor,
    xAxis: {
      interval: formData.xAxisInterval as AxisInterval,
      label: formData.xAxisLabel,
      tickLabelAngle: -Number(formData.xAxisTickLabelAngle),
    },
    hasY2Axis:
      yColumns.length > 1 &&
      formData.useY2Axis &&
      formData.layout === Layout.horizontal &&
      formData.queryMode !== QueryMode.raw,
    yAxis: {
      labelAngle: -Number(formData.yAxisLabelAngle ?? 0),
      labelAngle2: -Number(formData.y2AxisLabelAngle ?? 0),
      label: formData.yAxisLabel,
      tickLabelAngle: -Number(formData.yAxisTickLabelAngle),
      label2: formData.y2AxisLabel,
      tickLabelAngle2: -Number(formData.y2AxisTickLabelAngle),
    },
    data: resultData,
  };
  return result;
}
