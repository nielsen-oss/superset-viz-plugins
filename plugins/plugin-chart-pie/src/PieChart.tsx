/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React, { useState, FC, useEffect, memo, useRef, useCallback } from 'react';
import { styled, t, CategoricalColorNamespace } from '@superset-ui/core';
import {
  PieChart as RechartsPieChart,
  Pie as RechartsPie,
  Cell,
  RechartsFunction,
  Legend,
  PieProps as RechartsPieProps,
  LegendType,
} from 'recharts';
import { LegendPosition, renderActiveShape, getLegendProps, ActiveShapeProps } from './utils';

export const LABELS_MARGIN = 100;

type EventData = {
  color: string;
  id: string;
  type: LegendType;
  value: string;
};

type PieStylesProps = {
  height: number;
  width: number;
  legendPosition: LegendPosition;
};

type GroupBy<G extends string> = Record<G, string>;

export type PieChartData<G extends string, DK extends string> = GroupBy<G> & Record<DK, number>;

export type PieProps<G extends string = string, DK extends string = string> = {
  height: number;
  width: number;
  data: PieChartData<G, DK>[];
  dataKey: DK;
  isDonut?: boolean;
  onClick?: RechartsFunction;
  colorScheme: string;
  baseColor: string;
  legendPosition: LegendPosition;
  showLegend: boolean;
  showLabels: boolean;
  groupBy: G;
  labelType: string;
};

const Notification = styled.div`
  cursor: pointer;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: ${({ theme }) => theme.gridUnit * 4}px;
  border-radius: ${({ theme }) => theme.gridUnit * 2}px;
  color: ${({ theme }) => theme.colors.info.dark1};
  background-color: ${({ theme }) => theme.colors.info.light1};
`;

const Styles = styled.div<PieStylesProps>`
  border-radius: ${({ theme }) => theme.gridUnit * 2}px;
  height: ${({ height }) => height};
  width: ${({ width }) => width};
  overflow: auto;

  & .recharts-legend-item {
    cursor: pointer;
    white-space: nowrap;
  }
`;

const PieChart: FC<PieProps<string, string>> = memo(props => {
  const {
    dataKey,
    data,
    height,
    width,
    isDonut,
    colorScheme,
    showLegend,
    showLabels,
    groupBy,
    labelType,
    legendPosition,
  } = props;
  const [notification, setNotification] = useState<string | null>(null);
  const [activeIndex, setActiveIndex] = useState<number>(0);
  const [disabledDataKeys, setDisabledDataKeys] = useState<string[]>([]);
  const [legendWidth, setLegendWidth] = useState<number | null>(0);
  const [updater, setUpdater] = useState<number>(0);
  const rootRef = useRef<HTMLDivElement>(null);
  const isSideLegend =
    showLegend && (legendPosition === LegendPosition.right || legendPosition === LegendPosition.left);

  const currentData = data.filter(item => !disabledDataKeys.includes(item[groupBy]));
  const isExplore = window.location.pathname.includes('/explore');

  useEffect(() => {
    if (isSideLegend && rootRef.current) {
      const legend = rootRef.current.querySelector('.recharts-legend-wrapper');
      const currentWidth = legend?.getBoundingClientRect()?.width || null;
      if (currentWidth !== legendWidth) {
        setLegendWidth(currentWidth ? currentWidth + 20 : currentWidth);
      }
    }
  }, [updater, isSideLegend, legendWidth]);

  useEffect(() => {
    if (isSideLegend) {
      setLegendWidth(0);
    }
  }, [isSideLegend, props]);

  const forceUpdate = useCallback(() => setUpdater(Math.random()), []);

  useEffect(() => {
    forceUpdate();
  }, [forceUpdate, props]);

  const onPieEnter = (data: object, index: number) => setActiveIndex(index);
  const closeNotification = () => setNotification(null);

  const onClick = () => {
    if (isExplore) {
      setNotification(t('Sector was clicked, filter will be emitted on a dashboard'));
    }
  };

  const chartMargin = showLabels ? LABELS_MARGIN : 20;

  const handleLegendClick = ({ id }: EventData) => {
    let resultKeys = [];
    if (disabledDataKeys.includes(id)) {
      resultKeys = disabledDataKeys.filter(item => item !== id);
    } else {
      resultKeys = [...disabledDataKeys];
      resultKeys.push(id);
    }
    setDisabledDataKeys(resultKeys);
  };

  const chartHeight = height;
  const outerRadius = (width < chartHeight ? width : chartHeight) / 2 - chartMargin;
  const chartWidth =
    isSideLegend && legendWidth ? Math.max((outerRadius + chartMargin) * 2 + legendWidth, width) : width;

  const pieProps: RechartsPieProps & { key?: string | number } = {
    activeIndex,
    key: updater,
    data: currentData,
    dataKey,
    cx: isSideLegend ? outerRadius + chartMargin : '50%',
    outerRadius,
    label: showLabels
      ? labelProps => renderActiveShape({ ...labelProps, groupBy, labelType } as ActiveShapeProps)
      : false,
    onClick,
  };

  if (isDonut) {
    pieProps.activeShape = activeShapeProps =>
      renderActiveShape({ ...activeShapeProps, groupBy, labelType, isDonut: true });
    pieProps.onMouseEnter = onPieEnter;
    pieProps.innerRadius = outerRadius - outerRadius * 0.2;
    pieProps.label = false;
  }

  return (
    <Styles height={height} width={width} legendPosition={legendPosition} ref={rootRef}>
      {notification && <Notification onClick={closeNotification}>{notification}</Notification>}
      <RechartsPieChart key={updater} width={chartWidth} height={height}>
        {showLegend && (
          <Legend
            onClick={handleLegendClick}
            {...getLegendProps(legendPosition, height, legendWidth)}
            iconType="square"
            iconSize={10}
            payload={data.map(item => ({
              value: item[groupBy],
              id: item[groupBy],
              payload: item,
              type: disabledDataKeys.includes(item[groupBy]) ? 'line' : 'circle',
              color: CategoricalColorNamespace.getScale(colorScheme)(item[groupBy]),
            }))}
          />
        )}
        {((isSideLegend && legendWidth) || !isSideLegend) && (
          <RechartsPie {...pieProps}>
            {currentData?.map(entry => (
              <Cell fill={CategoricalColorNamespace.getScale(colorScheme)(entry[groupBy])} />
            ))}
          </RechartsPie>
        )}
      </RechartsPieChart>
    </Styles>
  );
});

export default PieChart;
