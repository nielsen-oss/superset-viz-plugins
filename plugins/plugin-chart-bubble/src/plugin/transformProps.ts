/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { ChartProps, DataRecord } from '@superset-ui/core';
import { Metric, SortingType } from '@superset-viz-plugins/plugin-chart-waterfall/src/plugin/utils';
import { LegendPosition } from '@superset-viz-plugins/plugin-chart-waterfall/src/types';

export type QueryData = {
  [key: string]: number | string;
};

type FormData = {
  xAxisColumn: string;
  xAxisLabel: string;
  yAxisLabel: string;
  yAxisLabelAngle: string;
  yAxisLogScale: boolean;
  xAxisTickLabelAngle: string;
  xAxisLogScale: boolean;
  numbersFormat: string;
  colorScheme: string;
  legendPosition: LegendPosition;
  maxNumberOfLegends: string;
  showGridLines: boolean;
  bubbleSize: boolean;
  entity: string;
  series: string;
  xAxis: { label: string } | string;
  yAxis: { label: string } | string;
  zAxis: { label: string } | string;
};

export default function transformProps(chartProps: ChartProps) {
  const { width, height, formData, queriesData } = chartProps;
  const data = queriesData[0].data as QueryData[];

  const {
    xAxisColumn,
    numbersFormat,
    legendPosition,
    maxNumberOfLegends,
    showGridLines,
    bubbleSize,
    xAxisLogScale,
    yAxisLogScale,
    colorScheme,
    entity,
    series,
  } = formData as FormData;

  return {
    xAxisDataKey: xAxisColumn,
    xAxisLabel: formData.xAxisLabel ?? '',
    yAxisLabel: formData.yAxisLabel ?? '',
    xAxisTickLabelAngle: -Number(formData.xAxisTickLabelAngle ?? 45),
    yAxisLabelAngle: -Number(formData.yAxisLabelAngle),
    xAxisLogScale,
    yAxisLogScale,
    width,
    height,
    bubbleSize: bubbleSize ?? 1000,
    legendPosition,
    maxNumberOfLegends,
    showGridLines,
    numbersFormat,
    colorScheme,
    data,
    entity,
    series,
    xAxis: formData.xAxis?.label ?? formData.xAxis,
    yAxis: formData.yAxis?.label ?? formData.yAxis,
    zAxis: formData.zAxis?.label ?? formData.zAxis,
  };
}
