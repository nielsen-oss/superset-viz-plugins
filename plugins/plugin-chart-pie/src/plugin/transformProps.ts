/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { ChartProps } from '@superset-ui/core';
import { PieChartData, PieProps } from '../PieChart';
import { LegendPosition } from '../utils';

type FormData<G extends string, DK extends string> = {
  colorScheme: string;
  isDonut: boolean;
  groupBy: G;
  metric: {
    label: DK;
  };
  colorPicker: string;
  showLegend: boolean;
  labelType: string;
  showLabels: boolean;
  legendPosition: LegendPosition;
};

export default function transformProps<G extends string, DK extends string>(chartProps: ChartProps): PieProps<G, DK> {
  const { width, height, formData, queryData, queriesData } = chartProps;
  const {
    colorScheme,
    isDonut,
    groupBy,
    metric,
    colorPicker,
    showLegend,
    labelType,
    showLabels,
    legendPosition,
  } = formData as FormData<G, DK>;
  const data = (queriesData?.[0]?.data || queryData.data) as PieChartData<G, DK>[];

  return {
    dataKey: metric.label,
    width,
    legendPosition,
    height,
    data: data.filter(item => item[metric.label] !== null).filter(item => item[groupBy] !== null),
    isDonut,
    baseColor: colorPicker,
    colorScheme,
    showLegend,
    showLabels: isDonut || showLabels,
    groupBy,
    labelType,
  };
}
