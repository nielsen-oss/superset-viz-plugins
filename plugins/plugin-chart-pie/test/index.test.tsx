/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import * as recharts from 'recharts';
import { supersetTheme, ThemeProvider } from '@superset-ui/core';
import PieChartPlugin from '../src';
import transformProps from '../src/plugin/transformProps';
import { categoryPercentageDonutNoLegend, legendLeftWithoutLabels, legendTopPercentage } from './__mocks__/pieProps';
import PieChart from '../src/PieChart';

jest.mock('recharts');

describe('plugin-chart-piw', () => {
  const RechartsPieChart = jest.fn(props => <div {...props} />);
  const Legend = jest.fn(() => <div />);
  const Pie = jest.fn(() => <div />);
  beforeEach(() => {
    // Recharts still have some UNSAFE react functions that failing test
    jest.spyOn(console, 'warn').mockImplementation(() => null);

    jest.clearAllMocks();

    // @ts-ignore
    recharts.PieChart = RechartsPieChart;
    // @ts-ignore
    recharts.Pie = Pie;
    // @ts-ignore
    recharts.Legend = Legend;
  });

  const getWrapper = (props: object) =>
    render(
      <ThemeProvider theme={supersetTheme}>
        {/*
       // @ts-ignore (no need emulate all props) */}
        <PieChart {...transformProps(props)} />
      </ThemeProvider>,
    );

  it('exists', () => {
    expect(PieChartPlugin).toBeDefined();
  });

  it('Chart with legend top / percentage labels ', () => {
    getWrapper(legendTopPercentage);
    expect({
      PieChartProps: RechartsPieChart.mock.calls[1],
      PieProps: Pie.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
    }).toMatchSnapshot();
  });

  it('Chart with legend left / without labels', () => {
    getWrapper(legendLeftWithoutLabels);
    expect({
      PieChartProps: RechartsPieChart.mock.calls[1],
      PieProps: Pie.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
    }).toMatchSnapshot();
  });

  it('Chart Donut with category and percentage labels / without legend', () => {
    getWrapper(categoryPercentageDonutNoLegend);
    expect({
      PieChartProps: RechartsPieChart.mock.calls[1],
      PieProps: Pie.mock.calls[1],
    }).toMatchSnapshot();
    expect(Legend).not.toBeCalled();
  });
});
