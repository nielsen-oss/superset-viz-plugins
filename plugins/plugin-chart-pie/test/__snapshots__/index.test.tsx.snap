// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-chart-piw Chart Donut with category and percentage labels / without legend 1`] = `
Object {
  "PieChartProps": Array [
    Object {
      "children": Array [
        false,
        <mockConstructor
          activeIndex={0}
          activeShape={[Function]}
          cx="50%"
          data={
            Array [
              Object {
                "AVG(efficiency)": 0.0566666666666667,
                "period": "2020",
              },
              Object {
                "AVG(efficiency)": 0.038,
                "period": "2018",
              },
              Object {
                "AVG(efficiency)": 0.0162068965517241,
                "period": "2019",
              },
              Object {
                "AVG(efficiency)": 0.0138888888888889,
                "period": "Total Period",
              },
              Object {
                "AVG(efficiency)": 0.011304347826087,
                "period": "2017",
              },
            ]
          }
          dataKey="AVG(efficiency)"
          innerRadius={80}
          label={false}
          onClick={[Function]}
          onMouseEnter={[Function]}
          outerRadius={100}
        >
          <Cell
            fill="#1f77b4"
          />
          <Cell
            fill="#ff7f0e"
          />
          <Cell
            fill="#2ca02c"
          />
          <Cell
            fill="#d62728"
          />
          <Cell
            fill="#9467bd"
          />
        </mockConstructor>,
      ],
      "height": 400,
      "width": 800,
    },
    Object {},
  ],
  "PieProps": Array [
    Object {
      "activeIndex": 0,
      "activeShape": [Function],
      "children": Array [
        <Cell
          fill="#1f77b4"
        />,
        <Cell
          fill="#ff7f0e"
        />,
        <Cell
          fill="#2ca02c"
        />,
        <Cell
          fill="#d62728"
        />,
        <Cell
          fill="#9467bd"
        />,
      ],
      "cx": "50%",
      "data": Array [
        Object {
          "AVG(efficiency)": 0.0566666666666667,
          "period": "2020",
        },
        Object {
          "AVG(efficiency)": 0.038,
          "period": "2018",
        },
        Object {
          "AVG(efficiency)": 0.0162068965517241,
          "period": "2019",
        },
        Object {
          "AVG(efficiency)": 0.0138888888888889,
          "period": "Total Period",
        },
        Object {
          "AVG(efficiency)": 0.011304347826087,
          "period": "2017",
        },
      ],
      "dataKey": "AVG(efficiency)",
      "innerRadius": 80,
      "label": false,
      "onClick": [Function],
      "onMouseEnter": [Function],
      "outerRadius": 100,
    },
    Object {},
  ],
}
`;

exports[`plugin-chart-piw Chart with legend left / without labels 1`] = `
Object {
  "LegendProps": Array [
    Object {
      "align": "left",
      "iconSize": 10,
      "iconType": "square",
      "layout": "vertical",
      "onClick": [Function],
      "payload": Array [
        Object {
          "color": "#1f77b4",
          "id": "2020",
          "payload": Object {
            "AVG(efficiency)": 0.0566666666666667,
            "period": "2020",
          },
          "type": "circle",
          "value": "2020",
        },
        Object {
          "color": "#ff7f0e",
          "id": "2018",
          "payload": Object {
            "AVG(efficiency)": 0.038,
            "period": "2018",
          },
          "type": "circle",
          "value": "2018",
        },
        Object {
          "color": "#2ca02c",
          "id": "2019",
          "payload": Object {
            "AVG(efficiency)": 0.0162068965517241,
            "period": "2019",
          },
          "type": "circle",
          "value": "2019",
        },
        Object {
          "color": "#d62728",
          "id": "Total Period",
          "payload": Object {
            "AVG(efficiency)": 0.0138888888888889,
            "period": "Total Period",
          },
          "type": "circle",
          "value": "Total Period",
        },
        Object {
          "color": "#9467bd",
          "id": "2017",
          "payload": Object {
            "AVG(efficiency)": 0.011304347826087,
            "period": "2017",
          },
          "type": "circle",
          "value": "2017",
        },
      ],
      "verticalAlign": "middle",
      "wrapperStyle": Object {
        "maxHeight": 400,
        "width": 0,
      },
    },
    Object {},
  ],
  "PieChartProps": Array [
    Object {
      "children": Array [
        <mockConstructor
          align="left"
          iconSize={10}
          iconType="square"
          layout="vertical"
          onClick={[Function]}
          payload={
            Array [
              Object {
                "color": "#1f77b4",
                "id": "2020",
                "payload": Object {
                  "AVG(efficiency)": 0.0566666666666667,
                  "period": "2020",
                },
                "type": "circle",
                "value": "2020",
              },
              Object {
                "color": "#ff7f0e",
                "id": "2018",
                "payload": Object {
                  "AVG(efficiency)": 0.038,
                  "period": "2018",
                },
                "type": "circle",
                "value": "2018",
              },
              Object {
                "color": "#2ca02c",
                "id": "2019",
                "payload": Object {
                  "AVG(efficiency)": 0.0162068965517241,
                  "period": "2019",
                },
                "type": "circle",
                "value": "2019",
              },
              Object {
                "color": "#d62728",
                "id": "Total Period",
                "payload": Object {
                  "AVG(efficiency)": 0.0138888888888889,
                  "period": "Total Period",
                },
                "type": "circle",
                "value": "Total Period",
              },
              Object {
                "color": "#9467bd",
                "id": "2017",
                "payload": Object {
                  "AVG(efficiency)": 0.011304347826087,
                  "period": "2017",
                },
                "type": "circle",
                "value": "2017",
              },
            ]
          }
          verticalAlign="middle"
          wrapperStyle={
            Object {
              "maxHeight": 400,
              "width": 0,
            }
          }
        />,
        false,
      ],
      "height": 400,
      "width": 800,
    },
    Object {},
  ],
  "PieProps": undefined,
}
`;

exports[`plugin-chart-piw Chart with legend top / percentage labels  1`] = `
Object {
  "LegendProps": Array [
    Object {
      "align": "center",
      "iconSize": 10,
      "iconType": "square",
      "layout": "horizontal",
      "onClick": [Function],
      "payload": Array [
        Object {
          "color": "#1f77b4",
          "id": "2020",
          "payload": Object {
            "AVG(efficiency)": 0.0566666666666667,
            "period": "2020",
          },
          "type": "circle",
          "value": "2020",
        },
        Object {
          "color": "#ff7f0e",
          "id": "2018",
          "payload": Object {
            "AVG(efficiency)": 0.038,
            "period": "2018",
          },
          "type": "circle",
          "value": "2018",
        },
        Object {
          "color": "#2ca02c",
          "id": "2019",
          "payload": Object {
            "AVG(efficiency)": 0.0162068965517241,
            "period": "2019",
          },
          "type": "circle",
          "value": "2019",
        },
        Object {
          "color": "#d62728",
          "id": "Total Period",
          "payload": Object {
            "AVG(efficiency)": 0.0138888888888889,
            "period": "Total Period",
          },
          "type": "circle",
          "value": "Total Period",
        },
        Object {
          "color": "#9467bd",
          "id": "2017",
          "payload": Object {
            "AVG(efficiency)": 0.011304347826087,
            "period": "2017",
          },
          "type": "circle",
          "value": "2017",
        },
      ],
      "verticalAlign": "top",
      "wrapperStyle": Object {
        "maxHeight": 400,
      },
    },
    Object {},
  ],
  "PieChartProps": Array [
    Object {
      "children": Array [
        <mockConstructor
          align="center"
          iconSize={10}
          iconType="square"
          layout="horizontal"
          onClick={[Function]}
          payload={
            Array [
              Object {
                "color": "#1f77b4",
                "id": "2020",
                "payload": Object {
                  "AVG(efficiency)": 0.0566666666666667,
                  "period": "2020",
                },
                "type": "circle",
                "value": "2020",
              },
              Object {
                "color": "#ff7f0e",
                "id": "2018",
                "payload": Object {
                  "AVG(efficiency)": 0.038,
                  "period": "2018",
                },
                "type": "circle",
                "value": "2018",
              },
              Object {
                "color": "#2ca02c",
                "id": "2019",
                "payload": Object {
                  "AVG(efficiency)": 0.0162068965517241,
                  "period": "2019",
                },
                "type": "circle",
                "value": "2019",
              },
              Object {
                "color": "#d62728",
                "id": "Total Period",
                "payload": Object {
                  "AVG(efficiency)": 0.0138888888888889,
                  "period": "Total Period",
                },
                "type": "circle",
                "value": "Total Period",
              },
              Object {
                "color": "#9467bd",
                "id": "2017",
                "payload": Object {
                  "AVG(efficiency)": 0.011304347826087,
                  "period": "2017",
                },
                "type": "circle",
                "value": "2017",
              },
            ]
          }
          verticalAlign="top"
          wrapperStyle={
            Object {
              "maxHeight": 400,
            }
          }
        />,
        <mockConstructor
          activeIndex={0}
          cx="50%"
          data={
            Array [
              Object {
                "AVG(efficiency)": 0.0566666666666667,
                "period": "2020",
              },
              Object {
                "AVG(efficiency)": 0.038,
                "period": "2018",
              },
              Object {
                "AVG(efficiency)": 0.0162068965517241,
                "period": "2019",
              },
              Object {
                "AVG(efficiency)": 0.0138888888888889,
                "period": "Total Period",
              },
              Object {
                "AVG(efficiency)": 0.011304347826087,
                "period": "2017",
              },
            ]
          }
          dataKey="AVG(efficiency)"
          label={[Function]}
          onClick={[Function]}
          outerRadius={100}
        >
          <Cell
            fill="#1f77b4"
          />
          <Cell
            fill="#ff7f0e"
          />
          <Cell
            fill="#2ca02c"
          />
          <Cell
            fill="#d62728"
          />
          <Cell
            fill="#9467bd"
          />
        </mockConstructor>,
      ],
      "height": 400,
      "width": 800,
    },
    Object {},
  ],
  "PieProps": Array [
    Object {
      "activeIndex": 0,
      "children": Array [
        <Cell
          fill="#1f77b4"
        />,
        <Cell
          fill="#ff7f0e"
        />,
        <Cell
          fill="#2ca02c"
        />,
        <Cell
          fill="#d62728"
        />,
        <Cell
          fill="#9467bd"
        />,
      ],
      "cx": "50%",
      "data": Array [
        Object {
          "AVG(efficiency)": 0.0566666666666667,
          "period": "2020",
        },
        Object {
          "AVG(efficiency)": 0.038,
          "period": "2018",
        },
        Object {
          "AVG(efficiency)": 0.0162068965517241,
          "period": "2019",
        },
        Object {
          "AVG(efficiency)": 0.0138888888888889,
          "period": "Total Period",
        },
        Object {
          "AVG(efficiency)": 0.011304347826087,
          "period": "2017",
        },
      ],
      "dataKey": "AVG(efficiency)",
      "label": [Function],
      "onClick": [Function],
      "outerRadius": 100,
    },
    Object {},
  ],
}
`;
